pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -179,42 +179,108 @@ def level_range(self):
 
 
 
-def finest_data(pdata, ilvl, hierarchy):
+def are_adjacent(lower_x, upper_x, atol=1e-6):
+    return np.abs(upper_x[0]-lower_x[-1]) < atol
+
+
+def overlap_mask(x, level, qty):
     """
-    given a FieldData, a levelnumber and a hierarchy
-    this function returns the coordinates and values
-    of the field at locations where no value exist on the
-    next finer level in the hierachy.
+    returns a mask for x where x is overlaped by
+    the qty patch datas on the specified level
     """
-    # although a priori doable for particles
-    # this just works for fields for now
 
-    assert pdata.quantity == 'field'
-    assert hierarchy.ndim == 1
+    is_overlaped = x != x #lgtm [py/comparison-of-identical-expressions]
+
+    # the strategy is to exclude from this patch data
+    # all nodes and values that are overlaped by next finer values
+    # the is_overlaped array will serve as a mask
+
+    for patch in level.patches:
+        pdata = patch.patch_datas[qty]
+        ghosts_nbr = pdata.ghosts_nbr[0]
+        finex = pdata.x[ghosts_nbr:-ghosts_nbr]
+        fine_dx = finex[1]-finex[0]
+        local_dx = x[1]-x[0]
+        if fine_dx < local_dx:
+            # the fine patch min and max are widened
+            # a bit so we include border nodes as overlaped
+            xmin, xmax = finex.min() - fine_dx/2, finex.max() + fine_dx/2
+            overlaped_idx = np.where( (x > xmin) & (x < xmax))[0]
+            is_overlaped[overlaped_idx] = True
+        else:
+            raise ValueError("level needs to have finer grid resolution than that of x")
+
+    return is_overlaped
+
+
+
+
+
+
+def finest_field(hierarchy, qty, time=None):
+    """returns a non-uniform contiguous (value,grid)
+       associated to the given hierarchy
+       the quantity "qty" must be a field
+    """
+    if time is None:
+        time = hierarchy.times()[0] # to replace with default
```

Comment:
can probably just rm these lines, hierarchy handles "time=None" internally

Reply:
you're probably right, I'll change it if that's true